File: WorldFeed/Client/src/app/core/Core/Client/Validators/TimelineClientRequiredParametersValidator.ts
import {ITimelineClientParametersValidator} from "./TimelineClientParametersValidator";
import {IUserQueryValidator} from "./UserQueryValidator";
import {TimelineParameters} from "./parameters-types";
import ArgumentNullException from 'src/app/c#-objects/TypeScript.NET-Core/packages/Core/source/Exceptions/ArgumentNullException';
import {IGetUserTimelineParameters} from "../../../Public/Parameters/TimelineClient/GetUserTimelineParameters";

export interface ITimelineClientRequiredParametersValidator extends ITimelineClientParametersValidator {
}

export class TimelineClientRequiredParametersValidator implements ITimelineClientRequiredParametersValidator {
  private readonly _userQueryValidator: IUserQueryValidator;

  constructor(userQueryValidator: IUserQueryValidator) {
    this._userQueryValidator = userQueryValidator;
  }

  public validate(parameters: TimelineParameters): void {
    if (parameters == null) {
      throw new ArgumentNullException(nameof(parameters));
    }

    if (this.isIGetUserTimelineParameters(parameters)) {
      this._userQueryValidator.throwIfUserCannotBeIdentified(parameters.User, `${nameof(parameters.User)}`);
    }
  }

  private isIGetUserTimelineParameters(parameters: TimelineParameters): parameters is IGetUserTimelineParameters {
    return (parameters as IGetUserTimelineParameters).customQueryParameters !== undefined;
  }
}

File: Sharebook/Server/src/Services/Post/Sharebook.Post.API/Client/Validators/TimelineClientRequiredParametersValidator.cs
namespace Sharebook.Post.Client.Validators
{
    using System;

    using Sharebook.Post.Application.Parameters.TimelineClient;

    public interface ITimelineClientRequiredParametersValidator : ITimelineClientParametersValidator
    {
    }

    public class TimelineClientRequiredParametersValidator : ITimelineClientRequiredParametersValidator
    {
        private readonly IUserQueryValidator userQueryValidator;

        public TimelineClientRequiredParametersValidator(IUserQueryValidator userQueryValidator)
        {
            this.userQueryValidator = userQueryValidator;
        }

        public void Validate(IGetHomeTimelineParameters parameters)
        {
            if (parameters == null)
            {
                throw new ArgumentNullException(nameof(parameters));
            }
        }

        public void Validate(IGetUserTimelineParameters parameters)
        {
            if (parameters == null)
            {
                throw new ArgumentNullException(nameof(parameters));
            }

            this.userQueryValidator.ThrowIfUserCannotBeIdentified(parameters.User, $"{nameof(parameters.User)}");
        }

        public void Validate(IGetMentionsTimelineParameters parameters)
        {
            if (parameters == null)
            {
                throw new ArgumentNullException(nameof(parameters));
            }
        }

        public void Validate(IGetRetweetsOfMeTimelineParameters parameters)
        {
            if (parameters == null)
            {
                throw new ArgumentNullException(nameof(parameters));
            }
        }
    }
}
